### PR TITLE
Add support for bgra8unorm-storage

### LINKED ISSUE #3359: Support `bgra8unorm-storage` extension
Prerequisite: https://github.com/gfx-rs/naga/issues/2195.

Metal supports this.

DX12 optionally supports this (see https://github.com/gpuweb/gpuweb/pull/3640#issuecomment-1341556918).
DX11 doesn't support this.
There is a possible workaround that we can implement if we want (see https://github.com/gpuweb/gpuweb/pull/3640#issuecomment-1340313220).

Vulkan optionally supports this if the format supports `STORAGE_WRITE_WITHOUT_FORMAT_BIT` (see https://github.com/KhronosGroup/Vulkan-Docs/issues/2027#issuecomment-1380608011).

### PR DESCRIPTION
**Pull request by nical**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Add change to CHANGELOG.md. See simple instructions inside file.

**Connections**

This is a continuation of the work that started in #3634 

Bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1829895

**Description**

Resolves #3359

**Testing**

There is a basic test in here.
The CTS also has [some coverage for this](https://github.com/gpuweb/cts/blob/8bb48275486acc5ce8ec25dcd3eabaeba80ed832/src/webgpu/api/validation/texture/bgra8unorm_storage.spec.ts#L45) although firefox hasn't updated to a version of the CTS that contains the test (but will soon).

### CODE REVIEW

**Comment by nical** [comment]
Currently blocked on https://github.com/gfx-rs/naga/pull/2550

File: wgpu-core/src/device/resource.rs
Comment (teoxoy) [inline wgpu-core/src/device/resource.rs]:
```suggestion
```

I was concerned we didn't have this for `rg11b10ufloat-renderable` as well but it turns out the code below takes care of these missing usages.

https://github.com/gfx-rs/wgpu/blob/b3135b9320820d34f214d0221b9cffbf0cdfe2c9/wgpu-core/src/device/resource.rs#L674-L687

File: wgpu-hal/src/vulkan/adapter.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs]:
```suggestion
        features2.contains(vk::FormatFeatureFlags::STORAGE_IMAGE)
            && features3.contains(vk::FormatFeatureFlags2::STORAGE_WRITE_WITHOUT_FORMAT)
```

File: wgpu-hal/src/vulkan/adapter.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs]:
```suggestion
        let mut properties3 = vk::FormatProperties3::default();
        let mut properties2 = vk::FormatProperties2::builder().push_next(&mut properties3);
```

Noticed ash has some nice utils we can use here. Note that the builder derefs to the type it's building.

File: tests/tests/bgra8unorm_storage.rs
Comment (teoxoy) [inline tests/tests/bgra8unorm_storage.rs:19]: Ran this test locally on my Nvidia dGPU on Vulkan (which supports the newly added `Features::BGRA8UNORM_STORAGE`) and it succeeded.

So, there should be no need for us to swizzle! 🎉 

Will update the naga issue with this info.

File: wgpu-hal/src/vulkan/adapter.rs
Comment (teoxoy) [inline wgpu-hal/src/vulkan/adapter.rs]:
```suggestion
    // This check gates the function call and structures used below.
    // TODO: check for (`VK_KHR_get_physical_device_properties2` or VK1.1) and (`VK_KHR_format_feature_flags2` or VK1.3).
    // Right now we only check for VK1.3.
    if api_version < vk::API_VERSION_1_3 {
        return false;
    }
```

**Review by teoxoy — CHANGES_REQUESTED** [review]

File: wgpu-hal/src/vulkan/adapter.rs
Comment (nical) [reply wgpu-hal/src/vulkan/adapter.rs]: Nice!

File: tests/tests/bgra8unorm_storage.rs
Comment (nical) [reply tests/tests/bgra8unorm_storage.rs:19]: yep it works on my amd laptop and on CI as well

Comment (teoxoy) [reply tests/tests/bgra8unorm_storage.rs:19]: I was afraid CI didn't support the feature so the test was skipped.

File: wgpu-types/src/lib.rs
Comment (teoxoy) [inline wgpu-types/src/lib.rs]:
```suggestion
```

File: player/tests/data/buffer-copy.ron
Comment (teoxoy) [inline player/tests/data/buffer-copy.ron]: I don't see why these had to change.

```suggestion
    features: 0x1_0000,
```

File: player/tests/data/clear-buffer-texture.ron
Comment (teoxoy) [inline player/tests/data/clear-buffer-texture.ron]:
```suggestion
    features: 0x0000_0004_0001_0000,
```

File: player/tests/data/zero-init-buffer.ron
Comment (teoxoy) [inline player/tests/data/zero-init-buffer.ron]:
```suggestion
    features: 0x1_0000,
```

**Review by teoxoy — COMMENTED** [review]
A few remaining things.

File: player/tests/data/buffer-copy.ron
Comment (nical) [reply player/tests/data/buffer-copy.ron]: @jinleili that change was from your initial PR, do you remember why these are needed ?

Comment (jinleili) [reply player/tests/data/buffer-copy.ron]: Just to pass the test, 0x1_0000 will incorrectly match the newly added feature

Comment (teoxoy) [reply player/tests/data/buffer-copy.ron]: I found https://github.com/gfx-rs/wgpu/pull/3023/commits/b924a4d534429f21eca15a9d6078d48fcdf76cf7 which indicates that it should correspond to `MAPPABLE_PRIMARY_BUFFERS`.
Which was changed in https://github.com/gfx-rs/wgpu/pull/3534 to be on bit 34.
```diff
-        const MAPPABLE_PRIMARY_BUFFERS = 1 << 16;
+        const MAPPABLE_PRIMARY_BUFFERS = 1 << 34;
```

File: wgpu-hal/src/vulkan/adapter.rs
Comment (teoxoy) [reply wgpu-hal/src/vulkan/adapter.rs]: Just a suggestion but note there is a typo (double "e") in `vk_1_3_reequirements`.

File: player/tests/data/buffer-copy.ron
Comment (teoxoy) [reply player/tests/data/buffer-copy.ron]: I opened https://github.com/gfx-rs/wgpu/pull/4232 to take care of these.

**Review by teoxoy — APPROVED** [review]
Looks good!

**Comment by teoxoy** [comment]
@nical the feature is missing from

https://github.com/gfx-rs/wgpu/blob/6c8ccdeaa9e01524aba51ba5ec3cc2ed8d1ab14d/deno_webgpu/01_webgpu.js#L5166

was brought up by @crowlKats in https://github.com/gfx-rs/wgpu/pull/3634#issuecomment-1495975447

**Comment by teoxoy** [comment]
I thought changes to deno require approval from @gfx-rs/deno but it seems my review is enough to land this. I remember this wasn't the case previously to adding the wgpu team to the codeowners file (https://github.com/gfx-rs/wgpu/pull/4162).

cc @cwfitzgerald

**Comment by crowlKats** [comment]
good catch, apologies for not having reviewed this.
Also, I believe previously PRs could be merged without my approval, its just that I was more attentive and actually left approvals or required changes relatively quickly

**Comment by nical** [comment]
Alright, this should be ready to land (note: I disabled the feature with vulkan on mac, because of the odd side effect causing the occlusion_query test to fail, I can't spend the time investigating that in the foreseeable future).

**Comment by teoxoy** [comment]
I reran the mac job that was failing https://github.com/gfx-rs/wgpu/actions/runs/6496021322/job/17672724374 and it ran successfully. I think it was just flakey (https://github.com/gfx-rs/wgpu/issues/4235#issuecomment-1761234759).

**Comment by nical** [comment]
Ok let's try one last time

**Comment by nical** [comment]
Ok it passed, let's merge.

**Comment by cwfitzgerald** [comment]
> I thought changes to deno require approval from https://github.com/orgs/gfx-rs/teams/deno but it seems my review is enough to land this. I remember this wasn't the case previously to adding the wgpu team to the codeowners file

@teoxoy wgpu has never required any reviews at all to land things, just CI passing - this is a policy I've kept to allow good-faith quick fixes without needing the delay of getting other people involved.

**Pull request merged**
